src/workers/layouts/nearestParentVertical.ts

```diff
@@ -0,0 +1,444 @@
+    upstreamRows: number[],
+    competingNodeId: string,
+  }
+  async function argueWithCompetingUpstreamPlacement({
+    desiredRow,
+    nodeStartX,
+    upstreamRows,
+    competingNodeId,
+  }: ArgueWithCompetingUpstreamPlacementProps): Promise<number> {
+    const competitor = layout.get(competingNodeId)
+
+    if (competitor === undefined) {
+      console.warn('NearestParentLayout - argueWithCompetingUpstreamPlacement: Competitor node was not found in the layout', competingNodeId)
+      return desiredRow
+    }
+
+    const [
+      competitorAboveConnections,
+      competitorBelowConnections,
+    ] = getNodeParentDirectionCounts(competingNodeId)
+    const nodeAboveConnections = upstreamRows.filter((upstreamPosition) => upstreamPosition < desiredRow).length
+    const nodeBelowConnections = upstreamRows.filter((upstreamPosition) => upstreamPosition > desiredRow).length
+
+    if (nodeAboveConnections > nodeBelowConnections) {
+      // node has more above
+      if (
+        competitorAboveConnections > competitorBelowConnections && competitorAboveConnections > nodeAboveConnections
+      ) {
+        // competitor has more above than below, and more above than node
+        // node wins, shove competitor up
+        await shove({
+          direction: -1,
+          nodeStartX,
+          desiredRow,
+        })
+        return desiredRow
+      }
+      if (competitorBelowConnections > competitorAboveConnections) {
+        // competitor has more below than above
+        // node wins, shove competitor down
+        await shove({
+          direction: 1,
+          nodeStartX,
+          desiredRow,
+        })
+        return desiredRow
+      }
+
+      // competitor has equal above and below, or node has more above
+      // place node above competitor
+      nodeShoveRecords.set(competingNodeId, -1)
+    }
+
+    if (nodeBelowConnections > nodeAboveConnections) {
+      // node has more below
+      if (
+        competitorBelowConnections > competitorAboveConnections && competitorBelowConnections > nodeBelowConnections
+      ) {
+        // competitor has more below than above, and more below than node
+        // node wins, shove competitor down
+        await shove({
+          direction: 1,
+          nodeStartX,
+          desiredRow,
+        })
+        return desiredRow
+      }
+
+      if (competitorAboveConnections > competitorBelowConnections) {
+        // competitor has more above than below
+        // node wins, shove competitor up
+        await shove({
+          direction: -1,
+          nodeStartX,
+          desiredRow,
+        })
+        return desiredRow
+      }
+
+      // competitor has equal above and below, or node has more below
+      // place node below competitor
+      nodeShoveRecords.set(competingNodeId, 1)
+    }
+
+    return await placeNearUpstreamNode(competingNodeId, nodeStartX)
+  }
+
+  function getNodeParentDirectionCounts(nodeId: string): [number, number] {
```

Comment:
nit: a named tuple would help with readability here
```suggestion
  function getNodeParentDirectionCounts(nodeId: string): [up: number, down: number] {
```